tidy3d/plugins/design/method.py

```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+
+
+class MethodGenAlg(MethodOptimise, ABC):
+    """A standard method for performing genetic algorithm search"""
+
+    # Args for the user
+    solutions_per_pop: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of solutions per population.",
+        description="TBD",
+    )
+
+    n_generations: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of generations.",
+        description="TBD",
+    )
+
+    n_parents_mating: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of parents mating.",
+        description="TBD",
+    )
+
+    stop_criteria: Optional[pd.constr(regex=r"\b(?:reach|saturate)_\d+\b")] = pd.Field(
+        title="Early stopping criteria.",
+        description="Define the early stopping criteria. Supported words are 'reach_X' or 'saturate_X' where X is a number. See PyGAD docs for more details",
+        default=None,
+    )
+
+    parent_selection_type: Optional[
+        Literal["sss", "rws", "sus", "rank", "random", "tournament"]
+    ] = pd.Field(
+        title="Parent selection type.",
+        description="TBD",
+        default="sss",
+    )
+
+    crossover_type: Optional[Literal["single_point", "two_points", "uniform", "scattered"]] = (
+        pd.Field(
+            title="Crossover type.",
+            description="TBD",
+            default="single_point",
+        )
+    )
+
+    crossover_prob: Optional[pd.confloat(ge=0, le=1)] = pd.Field(
+        title="Crossover probability.",
+        description="TBD",
+        default=0.8,
+    )
+
+    mutation_type: Optional[Literal["random", "swap", "inversion", "scramble", "adaptive"]] = (
+        pd.Field(
+            title="Crossover type.",
+            description="TBD",
+            default="random",
+        )
+    )
+
+    mutation_prob: Optional[pd.confloat(ge=0, le=1)] = pd.Field(
+        title="Crossover probability.",
+        description="TBD",
+        default=0.2,
+    )
+
+    # TODO: See if anyone is interested in having the full suite of PyGAD options - there's a lot!
+
+    def run(self, parameters: Tuple[ParameterType, ...], run_fn: Callable) -> Tuple[Any]:
+        """Defines the search algorithm for the GA"""
+
+        # Create fitness function combining pre and post fn with the tidy3d call
+        def fitness_function(ga_instance, solution, solution_idx):
+            # Break solution down to dict
+            sol_dict = self.sol_array_to_dict(solution, _param_keys, _param_converter)
+
+            # Iterate through solutions for batched and non-batched data
+            sol_out, aux_out = self._extract_output(run_fn(sol_dict))
+
+            self._flatten_and_append(aux_out, _store_aux)
+
+            return sol_out
+
+        def on_generation(ga_instance):
+            _store_parameters.append(ga_instance.population.copy())
+            _store_fitness.append(ga_instance.last_generation_fitness)
+            best_fitness = ga_instance.best_solution()[1]
+            print(
```

Comment:
for consistency, let's use `rich` similar to the web API. see `tidy3d/web/api/webapi.py` for examples on using `rich.Console`.

Reply:
Ah that's interesting. If I'm honest, I think there's later discussion needed to finalise what we will output to the user as there's a huge amount of metadata that can be retained for the optimisers that is currently ignored

Reply:
sounds good